lumigator/python/mzai/backend/backend/services/workflows.py

```diff
@@ -0,0 +1,166 @@
+
+from backend.services.datasets import DatasetService
+from backend.services.jobs import JobService
+
+
+class WorkflowService:
+    def __init__(
+        self,
+        job_service: JobService,
+        dataset_service: DatasetService,
+    ):
+        self._job_service = job_service
+        self._dataset_service = dataset_service
+
+    NON_TERMINAL_STATUS = [
+        JobStatus.CREATED.value,
+        JobStatus.PENDING.value,
+        JobStatus.RUNNING.value,
+    ]
+    """list: A list of non-terminal job statuses."""
+
+    # TODO: rely on https://github.com/ray-project/ray/blob/7c2a200ef84f17418666dad43017a82f782596a3/python/ray/dashboard/modules/job/common.py#L53
+    TERMINAL_STATUS = [JobStatus.FAILED.value, JobStatus.SUCCEEDED.value]
+    """list: A list of terminal job statuses."""
+
+    async def on_job_complete(self, job_id: UUID, task: Callable = None, *args):
+        """Watches a submitted job and, when it terminates successfully, runs a given task.
+
+        Inputs:
+        - job_id: the UUID of the job to watch
+        - task: the function to be called after the job completes successfully
+        - args: the arguments to be passed to the function `task()`
+        """
+        job_status = self._job_service.get_upstream_job_status(job_id)
+
+        loguru.logger.info(f"Watching {job_id}")
+        while job_status not in self.TERMINAL_STATUS and job_status in self.NON_TERMINAL_STATUS:
+            await asyncio.sleep(5)
+            job_status = self._job_service.get_upstream_job_status(job_id)
+
+        match job_status:
+            case JobStatus.FAILED.value:
+                loguru.logger.error(f"Job {job_id} failed: not running task {str(task)}")
+            case JobStatus.SUCCEEDED.value:
+                loguru.logger.info(f"Job {job_id} finished successfully.")
+                if task is not None:
+                    task(*args)
+            case _:
+                # NOTE: Consider raising an exception here as we *really* don't expect
+                # anything other than FAILED or SUCCEEDED.
+                loguru.logger.error(
+                    f"Job {job_id} has an unexpected status "
+                    f"({job_status}) that is not completed."
+                )
+
+    def _run_eval(
+        self,
+        inference_job_id: UUID,
+        request: WorkflowCreate,
+        background_tasks: BackgroundTasks,
+        experiment_id: UUID = None,
+    ):
+        # use the inference job id to recover the dataset record
+        dataset_record = self._dataset_service._get_dataset_record_by_job_id(inference_job_id)
+
+        # prepare the inputs for the evaluation job and pass the id of the new dataset
+        job_eval_dict = {
+            "name": f"{request.name}-evaluation",
+            "model": request.model,
+            "dataset": dataset_record.id,
+            "max_samples": request.max_samples,
+            "skip_inference": True,
```

Comment:
do we still need skip_inference here since we separate out inference and eval?

Reply:
Maybe not? Sorry to keep deferring stuff via my comments but this is existing code that I moved but didn't rewrite. I have thoughts about restructuring it but I would like to defer that to a future PR. This PR is focused on the re-organization effort to put things in locations to accomodate future work. I'm keeping all these comments in mind though so definitely good feedback for future!

Reply:
no worries!